xbmc/utils/Mp4ChplReader.cpp

```diff
@@ -0,0 +1,329 @@
+{
+  static_assert(std::is_unsigned_v<T>, "T must be an unsigned integer type");
+
+  T res = 0;
+  for (uint8_t byte : bytes)
+  {
+    res = (res << 8) | byte;
+  }
+  return res;
+}
+
+constexpr uint32_t Mkbetag(char a, char b, char c, char d)
+{
+  return (static_cast<uint32_t>(a) << 24) | (static_cast<uint32_t>(b) << 16) |
+         (static_cast<uint32_t>(c) << 8) | static_cast<uint32_t>(d);
+}
+
+// Structure to hold atom information
+struct AtomInfo
+{
+  uint32_t tag = 0;
+  int64_t size = 0;
+  int64_t startPos = 0;
+  int64_t endPos = 0;
+  int headerLength = 0;
+};
+
+ChplChapterResult ParseChpl(CFile& file, const int64_t chplSize, std::vector<ChplChapter>& out)
+{
+  // --- header ---
+  uint8_t version = 0;
+  if (file.Read(&version, 1) != 1)
+    return ErrorResult("Failed to read chpl version");
+
+  std::array<uint8_t, 3> flagBuf{};
+  if (file.Read(flagBuf.data(), flagBuf.size()) != static_cast<ssize_t>(flagBuf.size()))
+    return ErrorResult("Failed to read chpl flags");
+  const uint32_t flags =
+      (uint32_t(flagBuf[0]) << 16) | (uint32_t(flagBuf[1]) << 8) | uint32_t(flagBuf[2]);
+
+  uint32_t count = 0;
+  nanoseconds timeBase;
+
+  if (version == 0)
+  {
+    uint8_t tmp;
+    if (file.Read(&tmp, 1) != 1)
+      return ErrorResult("Failed to read No. of chapters in chpl version 0");
+    count = tmp;
+  }
+  else if (version == 1)
+  {
+    std::array<uint8_t, 4> timeBaseBuf{};
+    if (file.Read(timeBaseBuf.data(), timeBaseBuf.size()) !=
+        static_cast<ssize_t>(timeBaseBuf.size()))
+      return ErrorResult("Failed to read time base");
+    const uint32_t tbRaw = ReadBigEndian<uint32_t>(timeBaseBuf);
+    timeBase = (tbRaw != 0) ? nanoseconds(tbRaw) : V1_TIME_BASE;
+
+    std::array<uint8_t, 2> countBuf{};
+    if (file.Read(countBuf.data(), countBuf.size()) != static_cast<ssize_t>(countBuf.size()))
+      return ErrorResult("Failed to read No. of Chapters in chpl version 1");
+    count = (uint32_t(countBuf[1]) << 8) | countBuf[0];
+  }
+  else
+  {
+    const int64_t skip = chplSize - ATOM_HEADER_SIZE - CHPL_VERSION_FLAGS_SIZE;
+    if (skip <= 0 || file.Seek(skip, SEEK_CUR) != skip)
+      return ErrorResult("Failed to skip unknown chpl version");
+    return chplNone;
+  }
+
+  CLog::LogF(LOGDEBUG, "Found chpl atom: version={}, flags={}, containing {} chapters", version,
+             flags, count);
+
+  for (uint32_t i = 0; i < count; ++i)
+  { // Note: Spec claims 8 bytes, but actual implementations use 7 bytes
+    // Using 8 bytes will cause progressive data corruption in chapter titles and incorrect times
+    // See the example hex in xbmc/utils/Mp4ChplReader.h and breakdown for more information
+    std::array<uint8_t, 7> timestampBuf{};
+    if (file.Read(timestampBuf.data(), timestampBuf.size()) !=
+        static_cast<ssize_t>(timestampBuf.size()))
+      return ErrorResult("Failed to read timestamp of title");
+    const uint64_t ts = ReadBigEndian<uint64_t>(timestampBuf);
```

Comment:
(style) The indentation here is confusing. Adding braces to the `if` statement would fix this. Or perhaps a newline.

Sorry to bring so many nits, as it's a style choice it's entirely up to you. It shouldn't be much work to fix, but if it's legitimately a lot of work that you'd rather not do, just leave a quick justification and resolve this comment.

And if you do decide "won't fix", which is entirely your prerogative, just remember that braces and vertical whitespace can add code clarity for readers in the future.

Reply:
Entirely your call!